flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4069,27 +4116,27 @@ class ArrayExprLowering {
         operands.emplace_back([=](IterSpace) { return mlir::Value{}; });
       } else if (!argLowering) {
         // No argument lowering instruction, lower by value.
-        PushSemantics(ConstituentSemantics::RefTransparent);
-        auto lambda = genarr(*expr);
+        auto lambda =
+            genElementalArgumentAs(*expr, ConstituentSemantics::RefTransparent);
```

Comment:
nit: It looks like we could just push the semantics immediately here (as it was) and then call the new function (no need to pass a Semantics to it then), and the default fall-through case need not push any semantics. The initial check to see if a switch to DataValue is needed is the same but would use `this->semant` or `isReferentiallyOpaque()`, `isValueAttribute()`, etc.

Reply:
Yes, that is also a solution. I have updated the patch to do that.